0.5-style comprehensions

JeffBezanson opened the pull request:
This could be it! I no longer see a regression in 2d comprehensions. I also implemented @stevengj 's approach, using inference only in the empty case to avoid type inference regressions.

This also makes comprehensions shape-preserving, i.e. `[2x for x in rand(2,2)]` gives a 2d result.

Todo:
- [x] NEWS and docs
- [x] rebase after #16437
- [x] tests pass
- [ ] nanosoldier

JeffBezanson committed:
implement comprehensions as `collect` of a `Generator`

this removes `static_typeof` and `type_goto`

fixes #7258

JeffBezanson committed:
make `size` of a product iterator concatenate the sizes of its components

this removes an extra layer of wrapping with `IteratorND`

add some inline declarations for `Generator` iteration

remove unused `IteratorND` type

JeffBezanson committed:
use inferred type for empty comprehensions if it is a leaf type

this implements the "Steve Johnson compromise":
- non-empty comprehensions don't depend on inference at all
- in the empty case, we use either Union{} or the inferred type
- therefore there is no regression in type precision in cases where
  we can infer a leaf type

add Core.Inference.return_type

This computes a return type (at compile time if possible) in a way that handles
recursive dependencies in inference.

yuyichao commented:
> using inference only in the empty case to avoid type inference regressions.

So will there be another breakage in the next (or later) release cycle to return a different type (or raise an error) for empty case?

carnaval commented on base/inference.jl:
```diff
@@ -1786,39 +1721,6 @@ function typeinf_frame(frame)
-                end
-            end
-        end
-
```
:-)

tkelman commented:
> I no longer see a regression in 2d comprehensions.

What fixed this part? As of yesterday weren't you thinking you'd have to special-case 2d to avoid a regression?

JeffBezanson commented:
> So will there be another breakage in the next (or later) release cycle

I'm not sure. It might also be ok to give an error in the empty case. That's certainly safest, I just don't have a good handle on how important empty comprehensions are.

> What fixed this part?

Not sure; I haven't run the benchmark in 2 weeks. The generated code is definitely really different now. There are many more SROA variables.

vtjnash commented on test/arrayops.jl:
```diff
@@ -977,7 +977,7 @@ X = [ i+2j for i=1:5, j=1:5 ]
 @test X[4,5] == 14
 @test isequal(ones(2,3) * ones(2,3)', [3. 3.; 3. 3.])
-@test isequal([ [1,2] for i=1:2, : ], [1 2; 1 2])
+# @test isequal([ [1,2] for i=1:2, : ], [1 2; 1 2])
```
add a note that this (undocumented / unused / unknown) syntax is gone now?

JeffBezanson replied:
Yes, I was just going to bring that up for discussion. Unfortunately there's a bigger problem: this is hitting the infinite-recursion-via-codegen issue --- is there an issue for that? I was just trying to find it.

vtjnash replied:
I'm pretty sure it has an issue number, although I couldn't find it just now either.

JeffBezanson replied:
Found it: #16201. It was closed.

vtjnash commented on base/array.jl:
```diff
@@ -237,16 +237,29 @@ function _collect(cont, itr, ::HasEltype, isz::SizeUnknown)
+@eval function _default_eltype(itrt::ANY)
+    $(Expr(:meta, :pure))
+    try
+        rt = return_types(first, (itrt,))
```
`return_types` isn't pure and will occasionally return wrong answers if marked as such.

JeffBezanson replied:
Yes, I think we need a version that returns `Union{}` for incomplete frames --- I assume that's the issue?

vtjnash replied:
Yes, we have (well, had) it --- it was called static_typeof.... :/

JeffBezanson replied:
I hear you. But I think, at the very least, `return_type` is a better, more general interface to it.

vtjnash replied:
I would be happy for it to have a better interface. I think there's even some some valid uses of a static_typeof-like construct for optimizing closures (allowing the Boxes to be more accurately typed, but invisible to the user). I just don't think that `return_types` is capable of it, since you really need to return a type-inference edge, not a value, and we don't have first-class edges in type-inference. I'm not sure if you can have a first-class edge (it definitely permits breaking the monotonicity constraints on termination). It might be roughly correct for the edge to be represented through a closure?

```julia
return_types(f, at) do rt
    update!(?, rt)
end::Void
```

Alternatively, maybe you could return some sort of "promise" object that tracks the edge. I'm not really sure how that would look however.

JeffBezanson replied:
Yeah, it's tricky. Part of the difference between this and static_typeof is that this use is expected to be best-effort: we don't need *the* inferred type, just whatever type we can get, and if it's not a leaf type we ignore it anyway. Or we can just throw an error in the empty case.

vtjnash commented on base/array.jl:
```diff
@@ -237,16 +237,29 @@ function _collect(cont, itr, ::HasEltype, isz::SizeUnknown)
+        if length(rt) == 1 && isleaftype(rt[1])
+            return rt[1]
+        end
+    end
```
probably not a good idea to ignore all errors

JeffBezanson commented:
What to do with the little-acknowledged n-d comprehension syntax, `[ [1,2] for i=1:2, : ]`? I have a sense it's never used. Easiest to remove it, but I think we can implement it as a function now if necessary.

stevengj commented:
Is `[ [1,2] for i=1:2, : ]` even documented?  Seems better to remove it.

mschauer commented:
It prevents comprehensions becoming vectors of vectors which are awkward to use in julia (compared to arrays), so the functionality is useful.

GunnarFarneback commented:
Had I known about it I would have used it a couple of times when I have resorted to hcat([some comprehension producing a vector of vectors]...) kind of solutions.

stevengj commented:
You can always do `Vector{Int}[ [1,2] for i=1:2 ]`

GunnarFarneback commented:
Maybe I was unclear. I was talking of the situation of having a comprehension producing a vector of vectors similar to
```julia
julia> [[1, 2] for i = 1:3]
3-element Array{Array{Int64,1},1}:
 [1,2]
 [1,2]
 [1,2]
```
but actually wanting these collected in a matrix, which can be done with
```julia
julia> hcat([[1, 2] for i = 1:3]...)'
3x2 Array{Int64,2}:
 1  2
 1  2
 1  2
```
Had I known about the multidimensional comprehension syntax I would instead have done
```julia
julia> [[1, 2] for i = 1:3, :]
3x2 Array{Int64,2}:
 1  2
 1  2
 1  2
```
for a shorter solution, without splatting a possibly arbitrarily long vector. But I agree that the syntax is obscure and I would be as happy using a function instead.

nalimilan commented:
I didn't know about that syntax, but I've felt the need for an equivalent of `hcat(x...)` which doesn't involve splatting. Is there a solution currently?

andreasnoack commented:
I'm not aware of any so I ended up writing my own in
https://github.com/andreasnoack/TSVD.jl/blob/f777ff95213992c7b7f8daeeb9b4ae8b242f2fc0/src/common.jl#L1-L14

On Tue, May 31, 2016 at 4:54 PM, Milan Bouchet-Valat <
notifications@github.com> wrote:

> I didn't know about that syntax, but I've felt the need for an equivalent
> of hcat(x...) which doesn't involve splatting. Is there a solution
> currently?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/JuliaLang/julia/pull/16622#issuecomment-222817147>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAe0qeTZEZwnsVITUKkCVsHPg6268GGYks5qHJ_ygaJpZM4Io7vH>
> .
>

nalimilan commented:
Could be good to have something like this in Base.

mschauer commented:
Basically this can be covered by collecting a ``Concatenate``  iterator mentioned in #14805 which would take a generator as argument. The form
```
Array(concat((1:10 for i = 1:3)))
```
seems pretty clear and does not create unnecessary intermediates of the ranges in this case.

tkelman commented on src/julia-syntax.scm:
```diff
@@ -2016,88 +2016,43 @@
    (lambda (e)
-     (expand-forms (lower-comprehension #f       (cadr e) (cddr e))))
+     (if (any (lambda (x) (eq? x ':)) (cddr e))
+         (error "N-d comprehension syntax has been removed"))
```
only if some of the indices were colons, right?

JeffBezanson replied:
That's what `(if (any (lambda (x) (eq? x ':)) (cddr e))` checks.

tkelman replied:
I meant the error message should be more specific

JeffBezanson commented:
I believe the only remaining failure here is due to #16201. A bit odd, since the code that hits it (quadgk) doesn't seem to involve comprehensions, but somehow there's an interaction.

JeffBezanson commented:
Passing CI! Get excited!

vtjnash commented on base/inference.jl:
```diff
@@ -922,13 +919,23 @@ function abstract_apply(af::ANY, fargs, aargtypes::Vector{Any}, vtypes::VarTable
         end
     end
 
+    if f === return_type
```
Just adding a breadcrumb here that I'm still not sure if this is sound. Also, with the way this is currently being used, I expect it will tend to obscure any type errors, meaning this is probably very poorly tested right now.

vtjnash commented on base/inference.jl:
```diff
@@ -922,13 +919,23 @@ function abstract_apply(af::ANY, fargs, aargtypes::Vector{Any}, vtypes::VarTable
+            af = argtypes[2]
+            rt = abstract_call(isa(af,Const) ? af.val : af.parameters[1],
+                               (), Any[argtypes[2], tt.parameters[1].parameters...], vtypes, sv)
+            return Type{widenconst(rt)}
```
`Type{TypeVar(:R, widenconst(rt))}`

vtjnash commented on base/inference.jl:
```diff
@@ -3478,6 +3387,15 @@ function reindex_labels!(linfo::LambdaInfo, sv::InferenceState)
+function return_type(f::ANY, t::ANY)
+    rt = Union{}
+    for m in _methods(f, t, -1)
+        _, ty, _ = typeinf(m[3], m[1], m[2], false)
```
the third parameter `inferred` needs to be checked before using `ty`

JeffBezanson committed:
do some eta reduction (use `f` instead of `x->f(x)`) in Generator syntax

JeffBezanson committed:
take advantage of shape-preserving comprehensions in vectorize macros

vtjnash commented on base/inference.jl:
```diff
@@ -922,13 +919,32 @@ function abstract_apply(af::ANY, fargs, aargtypes::Vector{Any}, vtypes::VarTable
     end
 
+    if f === return_type
+        tt = argtypes[3]
```
`if length(argtypes) == 3`

vtjnash commented on base/inference.jl:
```diff
@@ -922,13 +919,32 @@ function abstract_apply(af::ANY, fargs, aargtypes::Vector{Any}, vtypes::VarTable
+        tt = argtypes[3]
+        if isType(tt)
+            af_argtype = tt.parameters[1]
+            if af_argtype <: Tuple
```
`&& isa(af_argtype, DataType)`

JeffBezanson commented:
Hrm, I distinctly remember doing doc changes for this and checking the box, but those changes seem to be gone now. Rebase mistake of some kind I suppose.

vtjnash commented on base/inference.jl:
```diff
@@ -922,13 +919,32 @@ function abstract_apply(af::ANY, fargs, aargtypes::Vector{Any}, vtypes::VarTable
+            if af_argtype <: Tuple
+                af = argtypes[2]
+                rt = abstract_call(isa(af,Const) ? af.val : af.parameters[1],
+                                   (), Any[argtypes[2], af_argtype.parameters...], vtypes, sv)
```
this will probably return `Union{}` first. maybe should special-case that below, since we don't really want to form `Type{TypeVar(:T, Union{})}` since that messes up the desired final answer

vtjnash commented on base/inference.jl:
```diff
@@ -3478,6 +3396,17 @@ function reindex_labels!(linfo::LambdaInfo, sv::InferenceState)
 end
 
+@eval function return_type(f::ANY, t::ANY)
+    $(Expr(:meta, :pure))
```
not pure.

JeffBezanson replied:
I still don't really get this. Even after we've inferred `Type{Int}` for `return_type(identity, Tuple{Int})`, we still have to actually call `return_type` at run time? That seems at best unnecessary, and at worst creates an opportunity for the run-time `return_type` to disagree.

vtjnash replied:
You can add a special-case to inlineable since we know our tfunc oracle is pure, but the function itself isn't pure (because `typeinf` can't be used from `@pure`)

JeffBezanson replied:
Ok, I can live with that. However, I thought the only reason `typeinf` can't be used from `@pure` is that inference might call a `@pure` function, but we already have a special case for that in pure_eval_call, effectively granting an exception for this function.

Keno commented:
Check your reflog, you should be able to find them if you ever committed them.

JeffBezanson committed:
doc/NEWS updates for comprehension changes

quinnj committed:
Tests and documentation for generators with guards and nesting.

JeffBezanson committed:
add comprehension filters (fixes #550) and nested loops (fixes #4867)

this also uses the new lowering for typed comprehensions, allowing all
comprehensions on unknown-length iterables (fixes #1457)

JeffBezanson commented:
Guards and nested ranges added.

tkelman commented on src/julia-syntax.scm:
```diff
@@ -1958,99 +1978,64 @@
+         (expand-forms `(comprehension (generator ,@(cdr e))))
+         (begin (if (and (eq? (caadr e) 'generator)
+                         (any (lambda (x) (eq? x ':)) (cddr (cadr e))))
+                    (error "N-d comprehension syntax has been removed"))
```
the error message should say "with `:`"

tkelman commented on NEWS.md:
```diff
@@ -60,6 +63,13 @@ Language changes
     implemented internally as `Tuple{Vararg{T,N}}` ([#11242]).
 
+  * Array comprehensions preserve the dimensions of the input ranges. For example,
+    `[ 2x for x in A]` will have the same dimensions as `A`.
```
this was done a while ago but should have a PR reference, right?

tkelman replied:
nevermind, looks like it should have this PR's number

tkelman replied:
still missing reference to this PR

JeffBezanson commented:
@nanosoldier `runbenchmarks(ALL, vs=:master)`

jrevels commented:
The value passed to `vs` has to be formatted as a string:

@nanosoldier `runbenchmarks(ALL, vs=":master")`

Nanosoldier's error reporting has a lot of room for improvement...

JeffBezanson commented:
Nanosoldier still seems to be having some trouble here.

jrevels commented:
Seems like this PR breaks JLD, which is used to load the benchmark parameters:

```
ERROR: LoadError: syntax: invalid dict comprehension
 in include_from_node1(::String) at ./loading.jl:426
 in macro expansion; at ./none:2 [inlined]
 in anonymous at ./<missing>:?
 in eval(::Module, ::Any) at ./boot.jl:234
 in process_options(::Base.JLOptions) at ./client.jl:243
 in _start() at ./client.jl:322
while loading /home/nanosoldier/.julia/v0.5/JLD/src/JLD.jl, in expression starting on line 1212
```

JeffBezanson commented:
Should be fixed now.

KristofferC commented:
@nanosoldier `runbenchmarks(ALL, vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/16622#issuecomment-229251598) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/6196bd90114b7cd527c537ae6a93b9b95722498c/e2d30c0_vs_cfc10f0/report.md). cc @jrevels

JeffBezanson commented:
This helps #6685 a bunch, since we now split out the comprehensions to a separate function.

davidagold commented:
@JeffBezanson on this branch, with `X` a `NullableArray{Float64, 1}`, `[ f(x) for x in X ]` appears to get parsed (is that the right word choice?) into `collect(Generator(f, X))`, which ends up returning an `Array`. I want to overload `collect` so that the aforementioned comprehension returns a `NullableArray`. Is the ability to do that part of your plan for this PR, or is it not really recommended?

yuyichao commented:
A comprehension always returns an `Array` (either on this branch or master). On 0.5, you can use something like `NullableArray(f(x) for x in X)` to do what you want.

davidagold commented:
That makes sense. Thanks, sorry for the noise.

JeffBezanson committed:
shorten call chain for `collect(::Generator)`

quinnj commented:
Looks like Appveyor was a 32-bit failure related to some kind of GitError. Needs a rebase.

JeffBezanson commented:
I've looked at some of the regressions identified by nanosoldier. One issue is #17342.

Another issue is that the comprehension body is not always inlined. If we want, that can (probably) be easily fixed by inserting an `:inline` node in the function the frontend generates.

Since comprehension expressions tend to be small, it's surprising that they might not be inlined. #17340 has an example of some unexpectedly-large IR from small code, when indexing a LinSpace.

JeffBezanson commented:
I think the causes of the (few) regressions here are well-understood and we'll be able to address them in the RC period, so I'll merge this now.
